File: src/SimpleLevelEditor.Formats/Types/EntityConfig/PointEntityVisualization.fs
namespace SimpleLevelEditor.Formats.Types.EntityConfig

open SimpleLevelEditor.Formats.Types

type PointEntityVisualization =
    | SimpleSphere    of Color: Rgb          * Radius: float32
    | BillboardSprite of TexturePath: string * Size: float32
    | Model           of ModelPath: string   * Size: float32

    member this.GetTypeId() : string =
        match this with
        | SimpleSphere _    -> nameof SimpleSphere
        | BillboardSprite _ -> nameof BillboardSprite
        | Model _           -> nameof Model

    member this.DeepCopy() : PointEntityVisualization =
        match this with
        | SimpleSphere (color, radius)        -> SimpleSphere (color, radius)
        | BillboardSprite (texturePath, size) -> BillboardSprite (texturePath, size)
        | Model (modelPath, size)             -> Model (modelPath, size)

File: src/SimpleLevelEditor.Formats.EntityConfig/PointEntityVisualization.cs
using Dunet;
using SimpleLevelEditor.Formats.Core;
using System.Diagnostics;
using System.Text.Json.Serialization;

namespace SimpleLevelEditor.Formats.EntityConfig;

[Union]
[JsonDerivedType(typeof(SimpleSphere), typeDiscriminator: nameof(SimpleSphere))]
[JsonDerivedType(typeof(BillboardSprite), typeDiscriminator: nameof(BillboardSprite))]
[JsonDerivedType(typeof(Model), typeDiscriminator: nameof(Model))]
public partial record PointEntityVisualization
{
	public sealed partial record SimpleSphere(Rgb Color, float Radius);
	public sealed partial record BillboardSprite(string TexturePath, float Size);
	public sealed partial record Model(string ModelPath, float Size);

	public string GetTypeId()
	{
		return this switch
		{
			SimpleSphere => nameof(SimpleSphere),
			BillboardSprite => nameof(BillboardSprite),
			Model => nameof(Model),
			_ => throw new UnreachableException(),
		};
	}

	public PointEntityVisualization DeepCopy()
	{
		return this switch
		{
			SimpleSphere simpleSphere => new SimpleSphere(simpleSphere.Color, simpleSphere.Radius),
			BillboardSprite billboardSprite => new BillboardSprite(billboardSprite.TexturePath, billboardSprite.Size),
			Model model => new Model(model.ModelPath, model.Size),
			_ => throw new UnreachableException(),
		};
	}
}
